File: files/2025-04-20/CSEC 20042025.R
library(tidyverse)
library(readxl)

path = "files/Challenge15-2025.xlsx"
input = read_excel(path, range = "B3:G5")
test = read_excel(path, range = "I3:K13")

result = input %>%
  rename(product = ...1) %>%
  pivot_longer(cols = -product, names_to = "month", values_to = "sales") %>%
  fill(sales, .direction = "down") %>%
  mutate(m_diff = sales - lag(sales, 1), .by = product) %>%
  replace_na(list(m_diff = 0)) %>%
  filter(m_diff > 0) %>%
  select(-sales) %>%
  uncount(m_diff) %>%
  mutate(Required = 1)

names(result) = names(test)

all.equal(result, test)
# [1] TRUE

File: files/2025-04-20/CSEC 20042025.py
import pandas as pd

path = "Challenge15-2025.xlsx"
input_data = pd.read_excel(path, usecols="B:G", skiprows=2, nrows=3)
test = pd.read_excel(path, usecols="I:K", skiprows=2, nrows=11)

result = (
    input_data.rename(columns={input_data.columns[0]: "product"})
    .melt(id_vars="product", var_name="month", value_name="sales")
    .sort_values(["product", "month"])
)
result["sales"] = result.groupby("product")["sales"].ffill()
result["m_diff"] = result.groupby("product")["sales"].diff().fillna(0)
result = result.loc[result["m_diff"] > 0, ["product", "month", "m_diff"]]
result = result.loc[result.index.repeat(result["m_diff"].astype(int))].reset_index(drop=True)
result["Required"] = 1
result.columns = test.columns

print(result.equals(test))
